Support character-wise visual mode word lookup

av3r4ge opened the pull request:
Support visual mode word lookup without overriding the unnamed
register.

beloglazov commented:
I didn't dig deep into it, but with this PR if you invoke the lookup from the visual mode twice, the whole content of the thesaurus buffer gets into the default register. Do you experience this as well?

av3r4ge commented:
Very interesting. And weird.
I experience it only when I lookup from visual mode and then lookup regularly.
Lookup twice or more by visual mode works well for me.

I'll dig into it and update the PR.

av3r4ge commented:
Ok. I've found and fixed the problem.
Actually it was always there, we just missed.

Because we use deletions in the Lookup function, it overrides the default register.
I fixed it by saving and restoring the register.

beloglazov commented on plugin/online-thesaurus.vim:
```diff
@@ -35,6 +35,7 @@ let s:path = shellescape(expand("<sfile>:p:h") . s:script_name)
 
 
 function! s:Lookup(word)
+    let reg_save = @@
```
Instead of saving the register here, we can just use the blackhole register on line 50 by replacing it with `1,$d _`

beloglazov commented on plugin/online-thesaurus.vim:
```diff
@@ -61,6 +62,21 @@ function! s:Lookup(word)
     exec 'resize ' . (line('$') - 1)
     setlocal nomodifiable filetype=thesaurus
     nnoremap <silent> <buffer> q :q<CR>
+    let @@ = reg_save
```
This becomes unnecessary.

beloglazov commented on plugin/online-thesaurus.vim:
```diff
@@ -61,6 +62,21 @@ function! s:Lookup(word)
+function! s:LookupVisual(mode)
+    " Only support character-wise visual mode
+    if a:mode == 'v'
+        let reg_save = @@
```
This still affects register `"*` in my setup. I guess it's better to use a named register like `@a`

beloglazov commented on plugin/online-thesaurus.vim:
```diff
@@ -61,6 +62,21 @@ function! s:Lookup(word)
+    if a:mode == 'v'
+        let reg_save = @@
+
+        silent exec ":normal! gvy"
```
This becomes `silent exec ':normal! gv"ay'`

beloglazov commented on plugin/online-thesaurus.vim:
```diff
@@ -61,6 +62,21 @@ function! s:Lookup(word)
+        silent exec ":normal! gvy"
+        call s:Lookup(@@)
+
+        let @@ = reg_save
```
`let @a = reg_save`

beloglazov commented on plugin/online-thesaurus.vim:
```diff
@@ -61,6 +62,21 @@ function! s:Lookup(word)
+        let reg_save = @@
+
+        silent exec ":normal! gvy"
+        call s:Lookup(@@)
```
`call s:Lookup(@a)`

beloglazov commented:
Sorry for taking so long to review your changes. Could you please address my comments? I'll then merge the PR.

av3r4ge commented:
Sorry for not fixing it yet. Busy time at work.
I'll take a look at it on the weekend.

beloglazov commented:
No worries, I have the same issue :)
